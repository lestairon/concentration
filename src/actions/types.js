export const CREATE_BOARD = "CREATE_BOARD";
export const FLIP_CARD = "FLIP_CARD";
export const TOGGLE_DISABLED = "TOGGLE_DISABLED";
export const SET_SOLVED = "SET_SOLVED";
export const INCREMENT_MOVES = "INCREMENT_MOVES";
export const RESET_FLIPPED = "RESET_FLIPPED";
export const TOGGLE_MENU = "TOGGLE_MENU";
