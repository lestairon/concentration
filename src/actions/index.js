import * as actions from "../constants/actionTypes";

const createBoard = ({ pairOfCards, ordered }) => ({
  type: actions.CREATE_BOARD,
  pairOfCards,
  ordered
});

const flipCard = id => ({
  type: actions.FLIP_CARD,
  id
});

const toggleDisabled = () => ({
  type: actions.TOGGLE_DISABLED
});

const setSolved = cards => ({
  type: actions.SET_SOLVED,
  cards
});

const incrementMoves = () => ({
  type: actions.INCREMENT_MOVES
});

const resetFlipped = () => ({
  type: actions.RESET_FLIPPED
});

const toggleMenu = () => ({
  type: actions.TOGGLE_MENU
});

const runTimer = () => {
  return (dispatch, getState) => {
    dispatch({ type: actions.RUN_TIMER, time: Date.now() });
    let timer = () => {
      if (getState().timer.running) {
        dispatch({ type: actions.TIME, time: Date.now() });
        setTimeout(timer, 1000);
      }
    };
    timer();
  };
};

const stopTimer = () => ({
  type: actions.STOP_TIMER
});

export {
  createBoard,
  toggleDisabled,
  setSolved,
  incrementMoves,
  toggleMenu,
  flipCard,
  resetFlipped,
  runTimer,
  stopTimer
};
