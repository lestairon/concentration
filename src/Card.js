import React from "react";
import { useSelector, useDispatch } from "react-redux";
import { flipCard } from "./actions";

const Card = ({ number, id, disabled }) => {
  const isFlipped = useSelector(
    ({ state }) => state.flippedCards.includes(id) || state.solved.includes(id)
  );
  const dispatch = useDispatch();

  const clickHandler = () => {
    if (!disabled && !isFlipped) dispatch(flipCard(id));
  };

  return (
    <div
      className="card"
      onClick={clickHandler}
      style={{
        background: isFlipped ? "#FFF" : "#8762b8ab",
        boxShadow: "0 4px 8px 0 rgba(0,0,0,0.2)"
      }}
    >
      {isFlipped ? number : ""}
    </div>
  );
};
export default Card;
