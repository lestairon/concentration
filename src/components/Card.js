import React from "react";
import { useSelector, useDispatch } from "react-redux";
import { flipCard, runTimer } from "../actions";
import PropTypes from "prop-types";
import { CardDiv } from "./components";

const Card = ({ number, id }) => {
  const isFlipped = useSelector(
    ({ boardState }) =>
      boardState.flippedCards.includes(id) || boardState.solved.includes(id)
  );
  const { disabled } = useSelector(({ boardState }) => boardState);
  const dispatch = useDispatch();
  const isRunning = useSelector(({ timer: { running } }) => running);

  const clickHandler = () => {
    if (!disabled && !isFlipped) dispatch(flipCard(id));
    !isRunning && dispatch(runTimer());
  };

  return (
    <CardDiv
      isFlipped={isFlipped}
      onClick={clickHandler}
      className={isFlipped && "active"}
    >
      {isFlipped && number}
    </CardDiv>
  );
};

Card.propTypes = {
  number: PropTypes.number.isRequired,
  id: PropTypes.number.isRequired
};

export default Card;
