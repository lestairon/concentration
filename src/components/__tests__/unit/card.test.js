import Card from "../../Card";
import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import { createStore } from "redux";
import rootReducer from "../../../reducers";
import "jest-styled-components";
import * as actions from "../../../actions";

const mockStore = createStore(rootReducer);

jest.spyOn(actions, "runTimer").mockReturnValue({ type: null });

const component = (number, id) =>
  render(
    <Provider store={mockStore}>
      <Card number={number} id={id} />
    </Provider>
  );

describe("card functionality", () => {
  it("card renders correctly", () => {
    const { container } = component(3, 0);
    expect(container).toMatchSnapshot();
  });

  it("flips the card on click", () => {
    const { container } = component(3, 0);
    expect(container.firstChild.textContent).toBe("");
    fireEvent.click(container.firstChild);
    expect(container.firstChild.textContent).toBe("3");
  });

  it("doesnt allow to unflip the card", () => {
    const { container } = component(3, 0);
    fireEvent.click(container.firstChild);
    fireEvent.click(container.firstChild);
    expect(container.firstChild.textContent).toBe("3");
  });
});
