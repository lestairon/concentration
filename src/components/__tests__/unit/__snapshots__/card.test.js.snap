// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`card functionality card renders correctly 1`] = `
.c0 {
  width: auto;
  height: auto;
  box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);
  font-family: Nunito,sans-serif;
  text-align: center;
  cursor: pointer;
  -webkit-transition: -webkit-transform 0.5s;
  -webkit-transition: transform 0.5s;
  transition: transform 0.5s;
  -webkit-transform-style: preserve-3d;
  -ms-transform-style: preserve-3d;
  transform-style: preserve-3d;
}

.c0,
.c0 .cardback,
.c0 .cardfront {
  border-radius: 10px;
}

.c0:hover,
.c0 > *:hover {
  -webkit-transform: scale(1.015);
  -ms-transform: scale(1.015);
  transform: scale(1.015);
}

.c0.active {
  -webkit-transform: rotateY(180deg);
  -ms-transform: rotateY(180deg);
  transform: rotateY(180deg);
}

.c0 .cardback,
.c0 .cardfront {
  position: absolute;
  width: 100%;
  height: 100%;
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
}

.c0 .cardfront {
  position: absolute;
  top: 0;
  left: 0;
}

.c0 .cardback {
  -webkit-transform: rotateY(180deg);
  -ms-transform: rotateY(180deg);
  transform: rotateY(180deg);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div>
  <div
    class="c0"
  >
    <img
      alt="Card"
      class="cardfront"
      draggable="false"
      src="education.svg"
    />
    <div
      class="cardback"
    />
  </div>
</div>
`;
